### PR TITLE
Kv symmetric key support 

### PR DESCRIPTION
**Pull request by g2vinay**
Enables Local Key support for AsyncKEK and KEK.

### CODE REVIEW

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java [hunk 1]
```diff
@@ -56,7 +56,7 @@
 public class CryptographyAsyncClient {
     static final String KEY_VAULT_SCOPE = "https://vault.azure.net/.default";
     static final String SECRETS_COLLECTION = "secrets";
-    JsonWebKey key;
+    AtomicReference<JsonWebKey> key;
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java]: What motivated your introduction of the `AtomicReference` type?

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java [hunk 2]
```diff
@@ -114,15 +116,15 @@ private void initializeCryptoClients() {
...
+        if (key.get().getKeyType().equals(RSA) || key.get().getKeyType().equals(RSA_HSM)) {
+            localKeyCryptographyClient = new RsaKeyCryptographyClient(key.get(), cryptographyServiceClient);
+        } else if (key.get().getKeyType().equals(EC) || key.get().getKeyType().equals(EC_HSM)) {
+            localKeyCryptographyClient = new EcKeyCryptographyClient(key.get(), cryptographyServiceClient);
+        } else if (key.get().getKeyType().equals(OCT)) {
+            localKeyCryptographyClient = new SymmetricKeyCryptographyClient(key.get(), cryptographyServiceClient);
         } else {
             throw logger.logExceptionAsError(new IllegalArgumentException(String.format(
-                "The Json Web Key Type: %s is not supported.", key.getKeyType().toString())));
+                "The Json Web Key Type: %s is not supported.", key.get().getKeyType().toString())));
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java]: This would be more readable with a `switch` statement

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyClientBuilder.java [hunk 3]
```diff
@@ -299,4 +300,18 @@ public CryptographyClientBuilder serviceVersion(CryptographyServiceVersion versi
...
+    /**
+     * Sets the jsonWebKey to be used for cryptography operations.
+     *
+     * <p>If {@code key} is provided then it takes precedence over key identifier and gets used for cryptography
+     * operations.</p>
+     *
+     * @param key The key to be used for cryptography operations.
+     * @return the updated builder object.
+     */
+    CryptographyClientBuilder key(KeyVaultKey key) {
```
[the remark this answers is not in the document]
Comment (JonathanGiles) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyClientBuilder.java]: It should just be 'key' - we leave the 'set' prefix off in builders. However, this method is package-private, so it is either Javadoc'd unnecessarily, or it should have been made public and been API reviewed, or it is only used internally and this is intentional - and should be commented as such to make that decision clear. From what I can see it is the last point - it is intended for internal use only.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/KeyEncryptionKeyClientBuilder.java [hunk 4]
```diff
@@ -114,6 +136,50 @@ public KeyEncryptionKey buildKeyEncryptionKey(String keyId) {
...
+    public Mono<? extends AsyncKeyEncryptionKey> buildAsyncKeyEncryptionKey(KeyVaultKey key) {
+        builder.key(key);
+        if (key == null) {
+            throw logger.logExceptionAsError(new IllegalStateException(
+                "Key is required to create key encryption key async client"));
+        }
+
+        if (key.getKey() == null) {
+            throw logger.logExceptionAsError(new IllegalStateException(
+                "Json Web key is required to create key encryption key async client"));
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/KeyEncryptionKeyClientBuilder.java]: These exceptions do not clearly state what the user should do to resolve the issue. More specificity means less confusion for end users.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java [hunk 5]
```diff
@@ -482,17 +484,17 @@ private void initializeCryptoClients() {
...
+            if (!available) {
+                return cryptographyServiceClient.unwrapKey(algorithm, encryptedKey, context);
+            }
 
-        if (!checkKeyPermissions(this.key.getKeyOps(), KeyOperation.WRAP_KEY)) {
-            return Mono.error(logger.logExceptionAsError(new UnsupportedOperationException(String.format("Unwrap Key Operation is not allowed "
-                + "for key with id %s", this.key.getId()))));
-        }
-        return localKeyCryptographyClient.unwrapKeyAsync(algorithm, encryptedKey, context, key);
+            if (!checkKeyPermissions(this.key.get().getKeyOps(), KeyOperation.WRAP_KEY)) {
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java]: yup, thanks for the catch.
Updated.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java [hunk 6]
```diff
@@ -625,25 +627,25 @@ private boolean checkKeyPermissions(List<KeyOperation> operations, KeyOperation
...
-                    this.key = keyVaultKey.getKey();
-                }
-                keyAvailableLocally = this.key.isValid();
-                initializeCryptoClients();
-            } catch (HttpResponseException | NullPointerException e) {
-                logger.warning("Failed to retrieve key from key vault");
-                logger.logExceptionAsWarning(e);
-                keyAvailableLocally = false;
+    private Mono<Boolean> ensureValidKeyAvailable() {
+        boolean keyAvailable = !(this.key == null && keyCollection != null);
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java]: Updated.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyClientBuilder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (g2vinay) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyClientBuilder.java]: Yeah, it is meant to be internal.
The javadoc is as good as a comment currently.
Will possibly exposed in future, when we decide to make this public across all languages.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/KeyVaultKey.java [hunk 7]
```diff
@@ -31,6 +31,35 @@
...
+        this.key = jsonWebKey;
+    }
+
+    /**
+     * Creates an instance of KeyVaultKey
+     * @param name the name of the key
+     * @param keyId the vault id name of the key
+     * @param jsonWebKey the json web key to be used for crypto operations
+     */
+    public KeyVaultKey(String name, String keyId, JsonWebKey jsonWebKey) {
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/KeyVaultKey.java]: Okay, that's true too.
Refactored this.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java (hunk 1 above)
Comment (g2vinay) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java]: On second thought, redesigned to keep it non atomic.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java (hunk 2 above)
Comment (g2vinay) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java]: Expandable String enum, cannot do that, as discussed.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/KeyEncryptionKeyClientBuilder.java (hunk 4 above)
Comment (g2vinay) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/KeyEncryptionKeyClientBuilder.java]: Updated the messages.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/KeyEncryptionKeyClientBuilder.java [hunk 8]
```diff
@@ -106,14 +128,58 @@ public KeyEncryptionKey buildKeyEncryptionKey(String keyId) {
...
+     * ({@link KeyEncryptionKeyClientBuilder#credential(TokenCredential) jsonWebKey vault credential} and
+     * {@code keyId} are required to build the {@link KeyEncryptionKeyAsyncClient async client}.</p>
+     *
+     * @param key the key to be used for crypto operations.
+     *
+     * @return A {@link KeyEncryptionKeyAsyncClient} with the options set from the builder.
+     * @throws IllegalStateException If {@link KeyEncryptionKeyClientBuilder#credential(TokenCredential)} or
+     * {@code key} have not been set.
+     */
+    public Mono<? extends AsyncKeyEncryptionKey> buildAsyncKeyEncryptionKey(KeyVaultKey key) {
```
Comment (srnagar) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/KeyEncryptionKeyClientBuilder.java]: Should this method be named `buildAsyncKeyEncryptionKeyClient`?

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/KeyVaultKey.java [hunk 9]
```diff
@@ -87,6 +97,30 @@ public KeyType getKeyType() {
         return key.getKeyOps();
     }
 
+    /**
+     * Creates an instance of KeyVaultKey
+     * @param keyId the identifier of the key
+     * @param jsonWebKey the json web key to be used for crypto operations
+     * @return The Key Vault Key object.
+     */
+    public static KeyVaultKey fromKeyId(String keyId, JsonWebKey jsonWebKey) {
```
Comment (srnagar) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/models/KeyVaultKey.java]: Good to have null checks on these helper methods.

File: sdk/keyvault/azure-security-keyvault-keys/src/test/resources/session-records/wrapUnwrapLocalSymmetricAK128.json [hunk 10]
```diff
@@ -0,0 +1,4 @@
+{
+  "networkCallRecords" : [ ],
```
Comment (srnagar) [inline sdk/keyvault/azure-security-keyvault-keys/src/test/resources/session-records/wrapUnwrapLocalSymmetricAK128.json]: Why do we need these empty network call records?

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java [hunk 11]
```diff
@@ -67,23 +66,25 @@
...
-        if (key.getKeyOps() == null) {
+        if (jsonWebKey.getKeyOps() == null) {
             throw new IllegalArgumentException("Json Web Key's key operations property is not configured");
         }
 
         if (key.getKeyType() == null) {
             throw new IllegalArgumentException("Json Web Key's key type property is not configured");
         }
-        this.key = key;
+        this.key = jsonWebKey;
```
Comment (schaabs) [inline sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java]: You probably want to store the entire KeyVaultKey so that you can validate the current time is in between notBefore expiresOn when performing encrypt, sign, and wapKey. But you could file an issue on this now and do a separate PR for it.

File: sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/KeyEncryptionKeyClientBuilder.java (hunk 8 above)
Comment (g2vinay) [reply sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/KeyEncryptionKeyClientBuilder.java]: This is used by other SDKs which utilize AsyncKEK and KEK.
So, from user perspective, they will use KV to build what other SDKs consume, 
this will be straight to the point and less confusing.
